include/aspect/postprocess/particle_crystal_preferred_orientation.h

```diff
@@ -0,0 +1,322 @@
+     * although this is not mandatory. It also allows other parts of the
+     * code to query the particles for information.
+     */
+    template <int dim>
+    class ParticleCrystalPreferredOrientation : public Interface<dim>, public ::aspect::SimulatorAccess<dim>
+    {
+      public:
+        /**
+         * Constructor.
+         */
+        ParticleCrystalPreferredOrientation();
+
+        /**
+         * Destructor.
+         */
+        ~ParticleCrystalPreferredOrientation();
+
+
+        /**
+         * Initialize function.
+         */
+        virtual void initialize ();
+
+        /**
+         * Execute this postprocessor. Derived classes will implement this
+         * function to do whatever they want to do to evaluate the solution at
+         * the current time step.
+         *
+         * @param[in,out] statistics An object that contains statistics that
+         * are collected throughout the simulation and that will be written to
+         * an output file at the end of each time step. Postprocessors may
+         * deposit data in these tables for later visualization or further
+         * processing.
+         *
+         * @return A pair of strings that will be printed to the screen after
+         * running the postprocessor in two columns; typically the first
+         * column contains a description of what the data is and the second
+         * contains a numerical value of this data. If there is nothing to
+         * print, simply return two empty strings.
+         */
+        virtual
+        std::pair<std::string,std::string> execute (TableHandler &statistics);
+
+        /**
+         * This funcion ensures that the particle postprocessor is run before
+         * this postprocessor.
+         */
+        virtual
+        std::list<std::string>
+        required_other_postprocessors () const;
+
+        /**
+        * Declare the parameters this class takes through input files.
+        */
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter file.
+         */
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+
+
+        /**
+         * Get volume weighted euler angles, using random draws to convert
+         * to a discrete number of orientations, weighted by volume.
+         */
+        std::vector<std::vector<double>> random_draw_volume_weighting(std::vector<double> fv,
+                                                                       std::vector<std::vector<double>> angles) const;
+
+        /**
+        * Wraps angle between 0 and 360 degrees.
+        */
+        double wrap_angle(const double angle) const;
+
+        /**
+         * Compute Z-X-Z euler angles from rotation matrix
+         */
+        std::vector<double> euler_angles_from_rotation_matrix(const Tensor<2,3> &rotation_matrix) const;
```

Comment:
I think this function and the one below would be good to have in utilities.cc. Maybe move there.